### PR TITLE
JAR Size Improvements

### PR DESCRIPTION
**Pull request by rhysdh540**
- minify all json files on assemble
- squished icons a little bit
- re-zip jar with highest compression settings

### CODE REVIEW

File: build.gradle.kts
Comment (embeddedt) [inline build.gradle.kts:297]: Does this map preserve order? I believe JAR files have stricter requirements around entry order than ZIP files.

File: build.gradle.kts
Comment (Nolij) [inline build.gradle.kts:313]: This is a nitpick but please replace `if(` with `if (`.

Also, should probably just check equality for `mcmod.info` instead of `endsWith`.

File: build.gradle.kts
Comment (Nolij) [inline build.gradle.kts:326]: I'd like to understand what the impact of this change is before merging it.

**Review by Nolij — CHANGES_REQUESTED** [review]
What is the current impact on output JAR size?

File: build.gradle.kts
Comment (rhysdh540) [reply build.gradle.kts:313]: `.endsWith` is necessary because `name` contains the whole path from the root of the jar. I'll change it to `File.separator + "mcmod.info"` just to be sure though.

File: build.gradle.kts
Comment (rhysdh540) [reply build.gradle.kts:297]: replaced with a LinkedHashMap

File: build.gradle.kts
Comment (rhysdh540) [reply build.gradle.kts:326]: more of a personal change I made, assemble just makes the jar, build makes it and tests it (and there are no tests here, so it's basically the same)
build depends on assemble so it functionally should have no difference if you are running build, but if you run just assemble then it works correctly

**Comment by rhysdh540** [comment]
> What is the current impact on output JAR size?

4,423 bytes smaller

**Comment by rhysdh540** [comment]
Note: if you *really* want a smaller jar, I can strip the class's LVTs and source files, which save another 17kb. This is technically a lossy change though (LVTs are used when throwing NullPointerExceptions, and the source files are used to give the file a class was compiled from in error messages)

**Comment by Nolij** [comment]
> Note: if you _really_ want a smaller jar, I can strip the class's LVTs and source files, which save another 17kb. This is technically a lossy change though (LVTs are used when throwing NullPointerExceptions, and the source files are used to give the file a class was compiled from in error messages)

For LVTs - what would the impact of this be if a NullPointerExpression was encountered?
For Source Files - if stacktraces would still list the class & method names, I see no issue with this for release builds.

If it's not too much effort, please make an easy way to not apply these "lossy" changes for test builds that would always apply for the `:publishMods` task. Maybe separate the compression logic in to a new Gradle task?

**Comment by rhysdh540** [comment]
LVTs:
consider:
```java
Object customVariableName = null
customVariableName.toString() // throws NPE
```
normally this would say something along the lines of "cannot invoke toString() because customVariableName is null)
with the LVT removed it'll just say the line (at Main.java:2)
and when decompiling, all the variables are replaced with auto-generated ones (var1, var2, etc)

for source files: my understanding is that they make inner classes name their outer class.

**Comment by Nolij** [comment]
> LVTs: consider:
> 
> ```java
> Object customVariableName = null
> customVariableName.toString() // throws NPE
> ```
> 
> normally this would say something along the lines of "cannot invoke toString() because customVariableName is null) with the LVT removed it'll just say the line (at Main.java:2) and when decompiling, all the variables are replaced with auto-generated ones (var1, var2, etc)
> 
> for source files: my understanding is that they make inner classes name their outer class when an exception is thrown in them

LVTs are not needed for release builds in that case.

For source files I'd like some clarification on this inner class behaviour but honestly assuming its scoped to release builds it should probably be fine.

File: build.gradle.kts
Comment (Nolij) [reply build.gradle.kts]: Oh, in that case the change is not necessary. Whichever way you prefer is fine here.

**Comment by rhysdh540** [comment]
ok after a little testing, stripping source files changes the file/line numbers for *everything* in zume in stacktraces to `Unknown Source`. Probably not the greatest idea in general but if you're confident that it won't throw anything then it does save a lot of space.

**Comment by Nolij** [comment]
> ok after a little testing, stripping source files changes the file/line numbers for _everything_ in zume in stacktraces to `Unknown Source`. Probably not the greatest idea in general but if you're confident that it won't throw anything then it does save a lot of space.

So it won't tell me the class or method name? Yeah that's non-ideal then... probably best to keep that information in that case.

**Comment by rhysdh540** [comment]
The answer is yes, unfortunately. I've kept it as an optional gradle property (`strip_source_files`) in case you want to turn it on though.

**Review by Nolij — COMMENTED** [review]
This one's a bit of a nit-pick. I can do these changes myself if you prefer, just reply and I'll mark it as resolved.

I'll test these changes this evening and approve assuming I don't find any issues.

**Comment by Nolij** [comment]
For some reason running the `:compressJar` task locally is resulting in everything except the jar manifest being removed...

**Review by Nolij — APPROVED** [review]

**Pull request merged**
